File: LC/Hard/127.py
class Solution(object):
    def ladderLength(self, beginWord, endWord, wordList):
        if endWord not in wordList:
            return 0
        n = len(beginWord)
        
        word_hash = defaultdict(list)
        for word in wordList:
            for j in range(n):
                pattern = word[:j] + "*" + word[j + 1:]
                word_hash[pattern].append(word)
            
        visited = {}
        q = [(beginWord, 1)]
        
        while q:
            word, level = q.pop(0)
            visited[word] = True
           
            for i in range(n):
                pattern = word[:i] + "*" + word[i + 1:]
                for nextWord in word_hash[pattern]:
                    if nextWord == endWord:
                        return level + 1
                    if nextWord not in visited:
                        q.append((nextWord, level + 1))
                word_hash[pattern] = []
                
        return 0           

File: LC/C++/Hard/127.cpp
class Solution {
public:
    int ladderLength(string beginWord, string endWord, vector<string>& wordList) {
        if (find(wordList.begin(), wordList.end(), endWord) == wordList.end()) {
            return 0;
        }

        std::unordered_map<std::string, vector<std::string>> hashes{};
        for (const std::string& word : wordList) {
            for (int i = 0; i < word.size(); ++i) {
                std::string hash = word.substr(0, i) + '*' + word.substr(i + 1);
                hashes[hash].push_back(word);
            }
        }

        std::queue<std::pair<int, std::string>> pq{};
        pq.push({1, beginWord});
        std::unordered_set<string> visited{};

        while (!pq.empty()) {
            auto [moves, curr] = pq.front();
            pq.pop();
            
            for (int i = 0; i < curr.size(); i++) {
                string pattern = curr.substr(0, i) + '*' + curr.substr(i + 1);

                for (const string& neigh : hashes[pattern]) {
                    if (curr == endWord) return moves;

                    if (visited.find(neigh) == visited.end()) {
                        pq.push({moves + 1, neigh});
                        visited.insert(neigh);
                    }
                }
            }
        }

        return 0;
    }
};